server/backends/migration_cache/migration_cache.go

```diff
@@ -232,3 +257,68 @@ func (mc *MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte)
 
 	return srcErr
 }
+
+type copyData struct {
+	d         *repb.Digest
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+}
+
+func (mc *MigrationCache) copyDataInBackground() error {
+	for {
+		select {
+		case <-mc.quitChan:
+			return nil
+		case c := <-mc.copyChan:
+			mc.copy(c)
+		}
+	}
+}
+
+func (mc *MigrationCache) copy(c *copyData) {
+	defer c.ctxCancel()
+
+	alreadyCopied, err := mc.dest.Contains(c.ctx, c.d)
+	if err != nil {
+		log.Warningf("Migration copy err, could not call Contains on dest cache: %s", err)
+		return
+	}
+
+	if !alreadyCopied {
```

Comment:
you can return early here if alreadyCopied is true

Reply:
Done